File: frontend-vue3/src/api/geo.ts
import client from './index'

export interface GeoDistributionParams {
  date: string
  lookback_days?: number
  level?: string
  top_n?: number
  segment_id?: number
}

export interface GeoDistributionItem {
  name: string
  user_count: number
  gmv: number
  // 后端返回的 "占比" 字段 = user_count / total_users * 100（百分比，非小数）
  gmv_ratio: number
  user_ratio: number
}

export interface GeoDistributionResponse {
  date: string
  level: string
  total_users: number
  total_gmv: number
  distribution: GeoDistributionItem[]
}

export interface GeoSegmentMatrixResponse {
  date: string
  matrix: Record<string, { province: string; user_count: number; gmv: number }[]>
  // color 来自后端 user_rfm 表的 segment_hex 字段
  segments: { id: number; name: string; color: string }[]
}

export interface GeoTrendResponse {
  time_points: string[]
  top_provinces: string[]
  trends: Record<string, number[]>
}

export function fetchGeoDistribution(params: GeoDistributionParams & { exclude_channels?: string[] }): Promise<GeoDistributionResponse> {
  return client.get('/v1/geo/distribution', { params })
}

export function fetchGeoSegmentMatrix(params: Omit<GeoDistributionParams, 'level' | 'top_n' | 'segment_id'> & { top_n?: number; exclude_channels?: string[] }): Promise<GeoSegmentMatrixResponse> {
  return client.get('/v1/geo/segment', { params })
}

export function fetchGeoTrend(params: { start_date: string; end_date: string; lookback_days?: number; top_n?: number; exclude_channels?: string[] }): Promise<GeoTrendResponse> {
  return client.get('/v1/geo/trend', { params })
}

File: backend/contracts/geo.py
"""芙清 CRM - Pydantic 契约模型"""
from __future__ import annotations
from typing import List, Any, Dict
from pydantic import BaseModel
from .types import RatioField, PercentageField, PpField  # Sprint 17 B2 全量 audit

class GeoDistributionItem(BaseModel):
    name: str
    user_count: int
    gmv: float
    # Sprint 17 B2 全量 audit: 2 个 0-1 decimal ratio 字段补 RatioField 标注
    user_ratio: "RatioField"
    gmv_ratio: "RatioField"


class GeoDistributionResponse(BaseModel):
    date: str
    level: str
    total_users: int
    total_gmv: float
    distribution: List[GeoDistributionItem]


class GeoSegmentMatrixResponse(BaseModel):
    date: str
    matrix: Dict[str, List[Dict[str, Any]]]
    segments: List[Dict[str, Any]]


class GeoTrendResponse(BaseModel):
    time_points: List[str]
    top_provinces: List[str]
    trends: Dict[str, Any]

